rflx/parser/session.py

```diff
@@ -0,0 +1,375 @@
+
+    string_literal = (
+        Literal('"').suppress() + Word(printables + " ", excludeChars='"') + Literal('"').suppress()
+    )
+    string_literal.setParseAction(lambda t: String(t[0]))
+
+    expr = Forward()
+
+    parameters = delimitedList(expr, delim=",")
+
+    lpar, rpar = map(Suppress, "()")
+
+    function_call = unqualified_identifier() + lpar + parameters + rpar
+    function_call.setParseAction(__parse_call)
+
+    conversion = qualified_identifier() + lpar + expr + rpar
+    conversion.setParseAction(__parse_conversion)
+
+    quantifier = (
+        Keyword("for").suppress()
+        - oneOf(["all", "some"])
+        + unqualified_identifier()
+        - Keyword("in").suppress()
+        + expr
+        - Keyword("=>").suppress()
+        + expr
+    )
+    quantifier.setParseAction(__parse_quantifier)
+
+    comprehension = (
+        Literal("[").suppress()
+        - Keyword("for").suppress()
+        + unqualified_identifier()
+        - Keyword("in").suppress()
+        + expr
+        - Keyword("=>").suppress()
+        + expr
+        + Optional(Keyword("when").suppress() + expr)
+        - Literal("]").suppress()
+    )
+    comprehension.setParseAction(__parse_comprehension)
+
+    null_message = Keyword("null") + Keyword("message")
+    null_message.setParseAction(lambda t: {})
+
+    components = delimitedList(
+        unqualified_identifier() + Keyword("=>").suppress() + expr, delim=","
+    )
+    components.setParseAction(lambda t: dict(zip(t[0::2], t[1::2])))
+
+    terms = delimitedList(unqualified_identifier() + Keyword("=").suppress() + expr, delim=",")
+    terms.setParseAction(lambda t: dict(zip(t[0::2], t[1::2])))
+
+    aggregate = (
+        qualified_identifier() + Literal("'").suppress() + lpar + (null_message | components) + rpar
+    )
+    aggregate.setParseAction(lambda t: MessageAggregate(t[0], t[1]))
+
+    atom = (
+        numeric_literal()
+        | bool_literal
+        | string_literal
+        | quantifier
+        | comprehension
+        | function_call
+        | conversion
+        | aggregate
+        | __variable()
+    )
+
+    attribute_designator = (
+        Keyword("Valid")
+        | Keyword("Present")
+        | Keyword("Length")
+        | Keyword("Head")
+        | Keyword("Opaque")
+    )
+
+    attribute = Literal("'").suppress() - attribute_designator
+    attribute.setParseAction(lambda t: (t[0], None))
+
+    field = Literal(".").suppress() - unqualified_identifier()
+    field.setParseAction(lambda t: ("Selected", t[0]))
+
+    binding = Keyword("where").suppress() + terms
+    binding.setParseAction(lambda t: ("Binding", t[0]))
+
+    suffix = binding ^ attribute ^ field
+
+    op_comp = Keyword("<") | Keyword(">") | Keyword("=") | Keyword("/=")
+
+    op_add_sub = Keyword("+") | Keyword("-")
+
+    op_mul_div = Keyword("*") | Keyword("/")
+
+    op_set = (Keyword("not") + Keyword("in")).setParseAction(lambda t: ["not in"]) | Keyword("in")
```

Comment:
It could, but I don't see why we should bother the user with an error in `not  in`. AFAICT the number of spaces is enforced nowhere else and IMHO it shouldn't be enforced here.

Reply:
> AFAICT the number of spaces is enforced nowhere else

We have no special handling for `end message`, `array of` and `null message`. Thus, in these cases we already enforce a single space. I think it's not a big burden for a user to not split these keywords in nonsensical ways.

Reply:
I think it would be consistent with the users expectations that tokens can be separated by an arbitrary amount of whitespace, as it's the case with Ada where we stole most of our syntax from. OTOH, it should be consistent with our other multi-component keywords and I don't have a strong option here. If you feel like making `not in` a single keyword, go ahead and change it.